### ISSUE TITLE
Question: How to upload image from client to nodal app?

### ISSUE THREAD

**Issue by pewh** [keithwhor/nodal#187]
It is a question rather than a issue.
Anybody know how to upload file from client to nodal app? Let's say I use form-data/multipart for upload image to server. Some tutorials I've read use express.js, so I can't figure out how to do this on nodal.

**Comment by keithwhor** [comment]
`this.params.buffer` on the controller contains the raw POST body data.

You should be able to read from that if you just send raw filedata.

**Comment by pewh** [comment]
I'm trying to parse formdata from buffer, so what I've done is:
```
      let formData
      const bufferStream = new stream.PassThrough();
      bufferStream.end(this.params.buffer);
      bufferStream.pipe(formData); // { name, file }
      console.log(formData);
```

but I got error: `TypeError: Cannot read property 'on' of undefined`

**Comment by keithwhor** [comment]
I can't help you right now, not familiar with the stream API you're using, can look into it soon.

**Comment by maxism** [comment]
@pewh there is an error in your code example. You can't just use pipe method to fill some variable with the data, because piping can be used in cases when you need to connect two or more specific objects (streams in terms of node.js) one into another. So your data flow will be looks like water running through the pipes.
In case of uploading images, you should use any module to parse the data from the HTTP request. For that purpose I am using Busboy to parse raw POST-data from Nodal's request buffer and then write image to the filesystem
```javascript
const Busboy = require('busboy');
const Stream = require('stream');
const Path = require('path');

...

let stream = new Stream.PassThrough();
let busboy = new Busboy({ headers: this._requestHeaders });

busboy.on('file', (fieldname, file, filename, encoding, mimetype) => {
  file.pipe(Fs.createWriteStream(Path.join(uploadFolderPath,filename)));
});

busboy.on('finish', () => {
  this.respond('Hooray');
});

stream.pipe(busboy);
stream.end(this.params.buffer);
```

**Comment by keithwhor** [comment]
This has been fixed in Nodal 0.10 :)

You can access images uploaded via a form by using `this.params.body.[FIELD]` where `[FIELD]` is your image from form data. :)
